urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/server/AbstractINDIServer.java

```diff
@@ -0,0 +1,812 @@
+   * Gets the port to which the Server listens.
+   *
+   * @return The port to which the Server listens.
+   */
+  protected int getListeningPort() {
+    return listeningPort;
+  }
+
+  /**
+   * The thread listens to the server socket and when a client connects, it is
+   * added to the list of clients.
+   */
+  @Override
+  public void run() {
+    try {
+      socket = new ServerSocket(listeningPort);
+    } catch (IOException e) {
+      System.err.println("Could not listen on port: " + listeningPort + " (maybe busy)");
+      return; // The thread will stop
+    }
+
+    System.err.println("Listening to port " + listeningPort);
+    System.err.flush();
+
+    mainThreadRunning = true;
+
+    while (mainThreadRunning) {
+      Socket clientSocket;
+
+      try {
+        clientSocket = socket.accept();
+      } catch (IOException e) {
+//   This is usually the escape point of the thread when the server is stopped.
+        System.err.println("Server has stopped listening to new Client connections");
+        mainThreadRunning = false;
+        return; // The thread will stop
+      }
+
+      if (clientSocket != null) {
+        if (acceptClient(clientSocket)) {
+          INDIClient client = new INDIClient(clientSocket, this);
+
+          clients.add(client);
+
+          connectionWithClientEstablished(client);
+        } else {
+          try {
+            clientSocket.close();
+          } catch (IOException e) {
+          }
+
+          // System.err.println("Client " + clientSocket.getInetAddress() + " rejected");
+        }
+      }
+    }
+  }
+
+  /**
+   * Gets if the server is listening for new Clients to connect.
+   *
+   * @return
+   * <code>true</code> if the server is listening for new Clients.
+   * <code>false</code> otherwise.
+   */
+  protected boolean isServerRunning() {
+    return mainThreadRunning;
+  }
+
+  /**
+   * Stops the server from listening new Clients. All connections with existing
+   * clients are also broken.
+   */
+  protected void stopServer() {
+    mainThreadRunning = false;
+    try {
+      socket.close();  // Close the socket in order to avoid accepting new connections
+    } catch (IOException e) {
+    }
+
+    for (int i = 0 ; i < clients.size() ; i++) {
+      if (clients.get(i) instanceof INDIClient) {
+        ((INDIClient) clients.get(i)).disconnect();
+      }
+    }
+  }
+
+  /**
+   * Loads all INDI for Java Drivers in a JAR file.
+   *
+   * @param jarFileName The jar file from which to load Drivers.
+   * @throws INDIException if there is a problem with the JAR file
+   */
+  public synchronized void loadJavaDriversFromJAR(String jarFileName) throws INDIException {
+    if (isAlreadyLoaded(jarFileName)) {
+      throw new INDIException("JAR file already loaded.");
+    }
+
+    ArrayList<String> list = getClassesInJAR(jarFileName);
+
+    File file = new File(jarFileName);
+    URL url;
+
+    try {
+      url = file.toURI().toURL();
+    } catch (MalformedURLException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 233.